RFE-1136 BZ#1796054 Infra node improvements and updates for new workflow

mburke5678 opened the pull request:
Taking over [this PR](https://github.com/openshift/openshift-docs/pull/32296/f) in Cody's absence

https://bugzilla.redhat.com/show_bug.cgi?id=1796054
https://issues.redhat.com/browse/RFE-1136

Previews:
- [Machine management - Creating infrastructure machine sets for production environments](https://deploy-preview-33057--osdocs.netlify.app/openshift-enterprise/latest/machine_management/creating-infrastructure-machinesets.html#creating-infrastructure-machinesets-production)
- [Post-installation - Creating infra machine sets](https://deploy-preview-33057--osdocs.netlify.app/openshift-enterprise/latest/post_installation_configuration/cluster-tasks.html#post-install-creating-infrastructure-machinesets)

mburke5678 commented:
@yuqi-zhang PTAL https://github.com/openshift/openshift-docs/pull/33057/commits/9dfd506453269d4f8525e828463f50bbb14f9a9c

mburke5678 commented:
@sferich888 PTAL https://github.com/openshift/openshift-docs/pull/33057/commits/9069b2a88353049c6de646a10887068e72f00e9d

sferich888 commented on modules/binding-infra-node-workloads-using-taints-tolerations.adoc:
```diff
@@ -1,19 +1,25 @@
 .Procedure
 
-. Use the following command to add a taint to the infra node to prevent scheduling user workloads on it:
+. If you have not configured a taint on your machine set, use the following command to add a taint to the infra node to prevent scheduling user workloads on it:
```
Should we show users how to validate that node's don't have taints?

mburke5678 replied:
@sferich888 The branch behins this PR is a clone of Cody's branch, with my changes. 

I added a sub-step to look in the node spec to see if there is a toleration. PTAL 
https://github.com/openshift/openshift-docs/pull/33057/commits/6d888fb5f5fc4839c90a11110a1f2efe78463617

sferich888 reviewed (approved):
/lgtm
This looks very similar of not exactly what @codyhoag had; so that's good. I do have a question that we may want to address; but other than that, it looks good.

mburke5678 commented:
@jhou1 PTAL

yuqi-zhang commented on modules/binding-infra-node-workloads-using-taints-tolerations.adoc:
```diff
@@ -1,20 +1,48 @@
 
+[IMPORTANT]
+====
+It is recommended to preserve the dual `infra,worker` label that is created for infra nodes and use taints and tolerations to manage nodes user workloads are scheduled on. If you remove the `worker` label from the node, you must create a custom pool to manage it; a node with a label other than `master` or `worker` is not recognized by the MCO without a custom pool. Maintaining the `worker` label  ...
```
Should this link to custom pool documentation

sferich888 replied:
@mburke5678 correct me if I am wrong but the way our docs; work we can't link to other sections of docs effectively.

mburke5678 replied:
Eric, that is correct. We are working out the details to get them back.

yuqi-zhang reviewed (approved):
Sorry for the late review, this mostly lgtm, one comment:

oarribas commented:
I think that the info in the following sections is contradictory:

https://github.com/openshift/openshift-docs/pull/33057/files#diff-adc76da643281f5cbdeb62b2b4b2fd87bb98101543d53b3a2b99400f07245188R11

https://github.com/openshift/openshift-docs/pull/33057/files#diff-2901799cc6024aefc0d385a7a42a24f967dc1d1e2b9a4a349b3529f39541e966R36


Is the recommendation to preserve the dual `infra,worker` label instead of using custom pools [1]?
I think that the custom-pools approach is widely used (and, as a side effect, it reduces the upgrade time as several nodes from different MCP are upgraded at the same time).


[1] https://github.com/openshift/machine-config-operator/blob/master/docs/custom-pools.md

yuqi-zhang commented:
Ah, good catch, that is a wording error. 

For full explanation,`infra,worker` label does leverage custom pools. Those should not be contradictory. Removing or not removing the extra `worker` label doesn't affect the MCO marking it as a infra node in the infra pool, since custom labels take priority.

yuqi-zhang commented on modules/binding-infra-node-workloads-using-taints-tolerations.adoc:
```diff
@@ -1,20 +1,48 @@
 
+[IMPORTANT]
+====
+It is recommended to preserve the dual `infra,worker` label that is created for infra nodes and use taints and tolerations to manage nodes user workloads are scheduled on. If you remove the `worker` label from the node, you must create a custom pool to manage it; a node with a label other than `master` or `worker` is not recognized by the MCO without a custom pool. Maintaining the `worker` label  ...
```
I just realized this is actually poorly phrased. `Maintaining the `worker` label allows the node to be managed by the default worker machine config pool` is ONLY correct if no custom pools that select the custom label exists. Any custom pool labelling overrides the worker label in terms of pool.

To break down what I mean:
1. `infra, worker` label + `infra` pool = node in infra pool
2. `infra` label + `infra` pool = node in infra pool
3. `infra,worker` label + NO `infra` pool = node in worker pool
4. `infra` label + NO `infra` pool = unmanaged node

We never want them to be in 4, so to be safe we are advertising 1/3

mburke5678 replied:
@yuqi-zhang  I changed to: _Maintaining the `worker` label allows the node to be managed by the default worker machine config pool, if no custom pools that select the custom label exists._

A user would maintain a `infra,worker` label in order to avoid an unmanaged node in the case that the infra pool goes away for some reason?

jhou1 commented:
@sunzhaohua2 @miyadav PTAL

sunzhaohua2 commented on modules/machineset-yaml-aws.adoc:
```diff
@@ -18,29 +30,63 @@ kind: MachineSet
+endif::infra[]
+ifdef::infra[]
+          node-role.kubernetes.io/infra: "" <3>      
+        taints: <4>
+          - key: node-role.kubernetes.io/infra
+            effect: NoSchedule
```
taints should be aligned with metadata
```
    spec:
      metadata:
        labels:
          node-role.kubernetes.io/infra: ""
      taints:
        - key: node-role.kubernetes.io/infra
          effect: NoSchedule
      providerSpec:
```

sunzhaohua2 commented on modules/machineset-yaml-azure.adoc:
```diff
@@ -18,29 +30,56 @@ kind: MachineSet
       creationTimestamp: null
       labels:
         machine.openshift.io/cluster-api-cluster: <infrastructureID> <1>
+        ifndef::infra[]
```
seems that this line is not needed, it looks like this from the preview doc
```
        machine.openshift.io/cluster-api-cluster: <infrastructureID> 
        ifndef::infra[]
        machine.openshift.io/cluster-api-machine-role: <role> 
        machine.openshift.io/cluster-api-machine-type: <role> 
        machine.openshift.io/cluster-api-machineset: <infrastructureID>-<role>-<region> 
        endif::infra[]
        ifdef::infra[]
        machine.openshift.io/cluster-api-machine-role: infra 
        machine.openshift.io/cluster-api-machine-type: infra 
        machine.openshift.io/cluster-api-machineset: <infrastructureID>-infra-<region> 
        endif::infra[]
```

sunzhaohua2 commented on modules/machineset-yaml-azure.adoc:
```diff
@@ -18,29 +30,56 @@ kind: MachineSet
       labels:
         machine.openshift.io/cluster-api-cluster: <infrastructureID> <1>
+        ifndef::infra[]
         machine.openshift.io/cluster-api-machine-role: <role> <2>
         machine.openshift.io/cluster-api-machine-type: <role> <2>
         machine.openshift.io/cluster-api-machineset: <infrastructureID>-<role>-<region> <3>
```
should be deleted

mikemckiernan replied:
Pointing out that this comment isn't marked as resolved.

sunzhaohua2 commented on modules/machineset-yaml-azure.adoc:
```diff
@@ -18,29 +30,56 @@ kind: MachineSet
         machine.openshift.io/cluster-api-machine-role: <role> <2>
         machine.openshift.io/cluster-api-machine-type: <role> <2>
         machine.openshift.io/cluster-api-machineset: <infrastructureID>-<role>-<region> <3>
+        endif::infra[]
+        ifdef::infra[]
```
should be deleted

sunzhaohua2 commented on modules/machineset-yaml-azure.adoc:
```diff
@@ -18,29 +30,56 @@ kind: MachineSet
+        machine.openshift.io/cluster-api-machine-role: infra <2>
+        machine.openshift.io/cluster-api-machine-type: infra <2>
+        machine.openshift.io/cluster-api-machineset: <infrastructureID>-infra-<region> <3>
+        endif::infra[]
```
should be deleted

sunzhaohua2 commented on modules/machineset-yaml-azure.adoc:
```diff
@@ -18,29 +30,56 @@ kind: MachineSet
+endif::infra[]
+ifdef::infra[]
+          node-role.kubernetes.io/infra: "" <2>      
+        taints: <4>
+        - key: node-role.kubernetes.io/infra
+          effect: NoSchedule
```
taints should be aligned with metadata
```
    spec:
      metadata:
        creationTimestamp: null
        labels:
          node-role.kubernetes.io/infra: ""
      taints:
      - key: node-role.kubernetes.io/infra
        effect: NoSchedule
      providerSpec:
```

sunzhaohua2 commented on modules/machineset-yaml-azure.adoc:
```diff
@@ -73,17 +112,37 @@ spec:
           sshPublicKey: ""
           subnet: <infrastructureID>-<role>-subnet <1> <2>
```
the subnet value does not seem to be a fixed value. Given infrastructureID is qeci-22538-spdj7, in some clusters it is like qeci-22538-worker-subnet, in some clusters it is like qeci-22538-spdj7-worker-subnet, we can add a comment, you can obtain the subnet by running the following command: 
```
  $  oc -n openshift-machine-api \
    -o jsonpath='{.spec.template.spec.providerSpec.value.subnet}{"\n"}' \
    get machineset/<infrastructureID>-worker-centralus1
```

sunzhaohua2 commented on modules/machineset-yaml-azure.adoc:
```diff
@@ -73,17 +112,37 @@ spec:
-            name: <role>-user-data <2>
+            name: worker-user-data <2>
           vmSize: Standard_D2s_v3
           vnet: <infrastructureID>-vnet <1>
```
the vnet value does not seem to be a fixed value. Given infrastructureID is qeci-22538-spdj7, in some clusters it is like qeci-22538-vnet,in some clusters it is like qeci-22538-spdj7-worker-vnet. we can add a comment, you can obtain the vnet by running the following command:

```
  $  oc -n openshift-machine-api \
    -o jsonpath='{.spec.template.spec.providerSpec.value.vnet}{"\n"}' \
    get machineset/<infrastructureID>-worker-centralus1
```

sunzhaohua2 commented on modules/machineset-yaml-gcp.adoc:
```diff
@@ -31,13 +43,27 @@ spec:
+endif::infra[]
+ifdef::infra[]
+          node-role.kubernetes.io/infra: "" <2>
+        taints: <3>
+        - key: node-role.kubernetes.io/infra
+          effect: NoSchedule
```
taints should be aligned with metadata

sunzhaohua2 commented on modules/machineset-yaml-gcp.adoc:
```diff
@@ -58,15 +89,25 @@ spec:
           networkInterfaces:
           - network: <infrastructureID>-network <1>
-            subnetwork: <infrastructureID>-<role>-subnet <2>
+            subnetwork: <infrastructureID>-<role>-subnet <1> <2>
```
should be `subnetwork: <infrastructureID>-worker-subnet`

mburke5678 replied:
@sunzhaohua2 Is this true for machine sets on Infra and GCP?

sunzhaohua2 replied:
@mburke5678 I think so, as there is no subnet `<infrastructureID>-infra-subnet ` in cluster,  if we set it like this, the machine will fail to create.

sunzhaohua2 commented on modules/machineset-yaml-gcp.adoc:
```diff
@@ -58,15 +89,25 @@ spec:
             - https://www.googleapis.com/auth/cloud-platform
           tags:
-          - <infrastructureID>-<role> <2>
+          - <infrastructureID>-<role> <1> <2>
```
should be <infrastructureID>-woker, If it is infra, the machine will fail to create

mburke5678 replied:
@sunzhaohua2 Is this true for machine sets on Infra and GCP?

sunzhaohua2 replied:
@mburke5678 yes, I think so.

sunzhaohua2 commented on modules/machineset-yaml-osp.adoc:
```diff
@@ -17,24 +30,47 @@ kind: MachineSet
+        machine.openshift.io/cluster-api-machine-role: infra <2>
+        machine.openshift.io/cluster-api-machine-type: infra <2>
+        machine.openshift.io/cluster-api-machineset: <infrastructure_ID>-infra <3>
+    spec:
```
seems forgot to add labels for osp machineset
```
      metadata:
        creationTimestamp: null
        labels:
          node-role.kubernetes.io/infra: "" 
```

sunzhaohua2 commented:
for modules/machineset-yaml-osp.adoc
line 109 should be` name: <infrastructure_ID>-worker`
line 111 should be `Name: <infrastructure_ID>-worker`
line 117 should be `name: worker-user-data`

mburke5678 commented:
@sunzhaohua2 Is [your comment here](https://github.com/openshift/openshift-docs/pull/33057#issuecomment-867427403) true for machine sets on Infra and GCP?

sunzhaohua2 commented:
> @sunzhaohua2 Is [your comment here](https://github.com/openshift/openshift-docs/pull/33057#issuecomment-867427403) true for machine sets on Infra and GCP?

yes, I tested these settings on gcp cluster, it works as expected. If we don't set it like this, machine will fail to create.

mikemckiernan commented on modules/binding-infra-node-workloads-using-taints-tolerations.adoc:
```diff
@@ -1,20 +1,48 @@
 +
+This example shows that the node has a taint. You can proceed with adding a toleration to your pod in the next step.
+
+.. If you have not configured a taint on your machine set, use the following command to add a taint to the infra node to prevent scheduling user workloads on it:
```
I noticed that you omitted "the following command" on line 22.  You could also omit here: "...machine set, add a taint..."

Up to you.

mikemckiernan commented on modules/creating-infra-machines.adoc:
```diff
@@ -31,6 +32,11 @@ spec:
++
+[NOTE]
+====
+Custom machine config pools inherit machine configs from the worker pool. Custom pools use any machine config targeted for the worker pool, but add the ability to also deploy changes only targeted at the custom pool. Since a custom pool inherits resources from the worker pool, any change to the worker pool also affects the custom pool.
```
nits and opinions:

* Would "...also deploy changes only targeted..." read more smoothly as "...also deploy changes that are only targeted..."?

* I tend to prefer "Because" instead of "Since".

mikemckiernan commented on modules/creating-infra-machines.adoc:
```diff
@@ -77,6 +83,13 @@ rendered-worker-daa08cc1e8f5fcdeba24de60cd955cc3            365c1cfd14de5b0e3b85
 +
+[NOTE]
+====
+Creating a machine config requires the associated nodes to reboot.
```
To newbs, this might be a little difficult to digest.  I'm not an MCO expert, so the foll is a *complete* guess:

After you create a machine config, the MCO reboots the nodes that are associated with the machine config to apply the configuration.

mikemckiernan commented on modules/creating-infra-machines.adoc:
```diff
@@ -104,9 +117,25 @@ spec:
 $ oc create -f infra.mc.yaml
 ----
+
+. Confirm that your new machine config pool is available.
```
End the instructional text with a colon.

mikemckiernan commented on modules/creating-infra-machines.adoc:
```diff
@@ -104,9 +117,25 @@ spec:
+[source,terminal]
+----
+NAME     CONFIG                                             UPDATED   UPDATING   DEGRADED   MACHINECOUNT
+infra    rendered-infra-b446e3cf9a1cc9be2c775fb742c41154    True      False      False      0
```
Complete ignoramus question: Is 0 typically the correct output for MACHINECOUNT at this point?

mburke5678 replied:
@yuqi-zhang Can you help with this question?  ^^

mburke5678 replied:
When you first create the MCP, the machine count is 0. Eventually, it comes up as 1.

yuqi-zhang replied:
It somewhat depends on your ordering of operations, application of machineconfigs don't affect this, only node labelling.

If you created the MCP first before scaling up infra nodes/moving workers to infra, it should be 0. If you labelled a node first and then create the MCP, it should almost immediately +1 infra -1 worker.

mikemckiernan commented on modules/machineset-yaml-aws.adoc:
```diff
@@ -4,12 +4,24 @@
+endif::[]
+ifeval::["{context}" == "post-install-cluster-tasks"]
+:infra:
+endif::[]
```
"Good idea fairy" sugg: Could you define an attribute that had the value `infra` in those two contexts, and if the attribute was not defined, the value is `<zone>`.  If yes, then the YAML that follows might be a little cleaner with `subs="attributes+"` and the attribute.

Completely up to you.

mikemckiernan commented on post_installation_configuration/cluster-tasks.adoc:
```diff
@@ -508,6 +508,20 @@ You can create a `MachineSet` object to host only infrastructure components. You
+
+After creating an infrastructure machine set, the `worker` and `infra` roles are applied to new infra nodes. Nodes with the `infra` role are not counted toward the total number of subscriptions that are required to run the environment, even when the `worker` role is also applied.
+
+However, with an infra node being assigned as a worker, there is a chance user workloads could get inadvertently assigned to an infra node. To avoid this, you can apply a taint to the infra node and tolerations for the pods you want to control.
```
Just a thought on the first sentence, up to you.

"However, when an infra node is assigned the `worker` role, there is a chance that user workloads can get assigned inadvertently to the infra node."

mikemckiernan commented on post_installation_configuration/cluster-tasks.adoc:
```diff
@@ -508,6 +508,20 @@ You can create a `MachineSet` object to host only infrastructure components. You
+.Additional resources
+
+* See xref:../nodes/scheduling/nodes-scheduler-about.adoc#nodes-scheduler-about[Controlling pod placement using the scheduler] for general information on scheduling a pod to a node.
+* See xref:moving-resources-to-infrastructure-machinesets[Moving resources to infrastructure machine sets] for instructions on scheduling pods to infra nodes.
```
Maybe "...for instructions about scheduling..."

I can't find guidance in the ISG about this, but if you search the ISG for "about ", then you'll notice that most instances are "for information about..." and they conscientiously did not use "information on."

mburke5678 replied:
It runs 3-2 in favor of _for information on_ in our docs, based on an Atom search. I will change it.

mikemckiernan reviewed (approved):
A few nitpicky items.  Looks good.

mburke5678 commented:
@sunzhaohua2 @yuqi-zhang I am sorry this is taking so long and so much effort. I couldn't get the process in the _Creating infrastructure machine sets_ topic to work. It seems to be lacking an explicit step to label the node you want as the infra node. It seems to assume you have already done this. I added a step to the _Creating a machine config pool for infrastructure machines_ topic to label the node. 
Also, the _Post-installation cluster tasks_ topic seems out of order. It suggests moving resources to the infra node before you create the infra node. I changed the structure to match what is in _Creating infrastructure machine sets_. 

Can I bother you to take a look at the new structures in this commit: https://github.com/openshift/openshift-docs/pull/33057/commits/4b2036e855094babcd39de4e7e51f474bb7635e3

I will get previews in a bit.

Current structures: [
Post-installation cluster tasks](https://docs.openshift.com/container-platform/4.7/post_installation_configuration/cluster-tasks.html)
[Creating infrastructure machine sets](https://docs.openshift.com/container-platform/4.7/machine_management/creating-infrastructure-machinesets.html)

yuqi-zhang commented on modules/creating-infra-machines.adoc:
```diff
@@ -74,9 +94,18 @@ rendered-worker-4f110718fe88e5f349987854a1147755            365c1cfd14de5b0e3b85
 +
+[NOTE]
+====
+After you create a machine config, the MCO reboots the nodes that are associated with the machine config to apply the configuration.
```
`After you create a machine config, the MCO reboots the nodes that are associated with the machine config to apply the configuration.` -> `After the new MCP is created, the MCO generates a new rendered config for that pool, and associated nodes of that pool will reboot to apply the new configuration`

yuqi-zhang commented on modules/machineset-creating.adoc:
```diff
@@ -177,8 +180,11 @@ status:
 ----
+////
+////
+Thi step is not needed. I have not labeled a node yet.
```
Is this meant for the doc?

yuqi-zhang commented on modules/machineset-creating.adoc:
```diff
@@ -129,6 +129,9 @@ endif::win[]
 When the new machine set is available, the `DESIRED` and `CURRENT` values match. If the machine set is not available, wait a few minutes and run the command again.
 
+////
+This step is not needed. No machines at this point.
```
Is this meant for the doc?

yuqi-zhang reviewed (commented):
A few comments, I can't really see without the preview, but so long as somewhere references how to add nodes to the infra pool, I am ok with it

mburke5678 commented:
/cherrypick enterprise-4.5

mburke5678 commented:
/cherrypick enterprise-4.7

mburke5678 commented:
/cherrypick enterprise-4.8
